AWS Bedrock support

kaeawc opened the pull request:
Wrote this to fulfill https://github.com/JetBrains/koog/issues/162

Adds model support for the current slate of available Bedrock models with supporting unit tests and MCP tool calling. I'm happy to add more documentation than just the kdoc, let me know what's appropriate here.

---

#### Type of the change
- [x] New feature
- [ ] Bug fix
- [ ] Documentation fix

#### Checklist for all pull requests
- [x] The pull request has a description of the proposed change
- [x] I read the [Contributing Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md) before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [x] Tests for the changes have been added
- [x] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [x] An issue describing the proposed change exists
- [x] The pull request includes a link to the issue
- [x] The change was discussed and approved in the issue
- [x] Docs have been added / updated

aozherelyeva commented on examples/src/main/kotlin/ai/koog/agents/example/mcp/GoogleMapsMcpClient.kt:
```diff
@@ -58,6 +60,21 @@ fun main() {
+
+        val bedrockAgent = AIAgent(
+            executor = simpleBedrockExecutor(awsAccessKey, awsSecretKey),
+            systemPrompt = "You are a helpful assistant. Answer user questions concisely using AWS Bedrock.",
```
Nit: this may be extensive information to include in the system prompt. As `simpleBedrockExecutor` is an instance of `SingleLLMPromptExecutor`, the agent won't be able to use any other executor anyways.

sdubov commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jvmMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.jvm.kt:
```diff
@@ -0,0 +1,645 @@
+                is Message.User -> {
+                    // Check for image content if present
+                    if (msg.mediaContent.isNotEmpty()) {
+                        require(model.capabilities.contains(LLMCapability.Vision.Image)) {
```
@kaeawc could you please clarify this check for me? As far as I understand, the model can return some media type, that is not an image, but we explicitly check for Image capabilities here. Please correct me if I miss anything.

sdubov commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jvmMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.jvm.kt:
```diff
@@ -0,0 +1,645 @@
+                            msg.mediaContent.forEach { media ->
+                                when (media) {
+                                    is MediaContent.Image -> {
+                                        val imageData = if (media.isUrl()) {
```
It seems that `imageData` variable is unused after creating. Could you please clarify if it is used anywhere below?

sdubov commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jsMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.js.kt:
```diff
@@ -0,0 +1,42 @@
+    // JS specific: No actual credentials needed as it will be unsupported.
+}
+
+public actual fun createBedrockLLMClient(
```
Could you please add KDoc to the public function?

sdubov commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jsMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.js.kt:
```diff
@@ -0,0 +1,42 @@
+
+    override fun executeStreaming(prompt: Prompt, model: LLModel): Flow<String> {
+        console.warn(unsupportedMessage)
+        return flowOf()
```
Could you please clarify the difference in the behavior between the `execute` (throw) and the `executeStreaming` (return empty flow) methods? Would it be more convenient to throw in both cases?

sdubov commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jvmTest/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClientTest.kt:
```diff
@@ -0,0 +1,511 @@
+import kotlin.test.assertContains
+import kotlin.test.assertFails
+
+class BedrockLLMClientTest {
```
Nice! Thank you for adding all those tests :+1:

kaeawc commented:
@sdubov thanks for the very thorough and thoughtful review! I've addressed all feedback and made sure tests are still passing. I rebased from latest `develop` to resolve merge conflicts. Let me know if there's anything else you'd like me to address.

kaeawc commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jsMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.js.kt:
```diff
@@ -0,0 +1,52 @@
+    clock: Clock
+): LLMClient = BedrockLLMClientJS()
+
+private class BedrockLLMClientJS : LLMClient {
+    private val unsupportedMessage = "AWS Bedrock client is not supported on Kotlin/JS"
+
+    override suspend fun execute(
+        prompt: Prompt,
+        model: LLModel,
+        tools: List<ToolDescriptor>
+    ): List<Message.Response> {
+        console.warn(unsupportedMessage)
+        throw UnsupportedOperationException(unsupportedMessage)
+    }
+
+    override fun executeStreaming(prompt: Prompt, model: LLModel): Flow<String> {
+        console.warn(unsupportedMessage)
+        throw UnsupportedOperationException(unsupportedMessage)
+    }
+}
```
Sure thing!

kaeawc commented:
@EugeneTheDev thanks for the thoughtful review! I've done most of the changes requested but have a question:

The last part have left to finish is using data classes instead of json objects. I was wondering if we could reuse the data classes that are already in koog for providers like Anthropic since those objects should be exactly the same. That might help with maintenance overall. Alternatively, if y'all don't want those data classes to be a public APl on those Gradle modules I could refactor them into separate serialization modules? But don't know if it's a concern at all.

Let me know what you think, happy to do what works best for koog.

kaeawc commented:
@EugeneTheDev @Ololoshechkin let me know if there's anything else you'd like me to change or update. The serialization data class change was a pretty big refactor, I went through the AWS model lists and available documentation for the current models. I also scoped down the number of providers supported out of the gate so it would be easier to get this out.

EugeneTheDev commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jvmMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.kt:
```diff
@@ -0,0 +1,292 @@
+
+        val requestBody = when (modelFamily) {
+            is BedrockModelFamilies.AI21Jamba ->
+                if (model.id.startsWith("ai21.jamba")) {
```
Looks like these if's here and below are redundant, since this check is already performed in `getBedrockModelFamily`

EugeneTheDev commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jvmMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.kt:
```diff
@@ -0,0 +1,292 @@
+                BedrockAI21JambaSerialization.createJambaRequest(prompt, model, emptyList())
+            )
+            is BedrockModelFamilies.AmazonNova -> {
+                if (model.id.startsWith("amazon.nova")) {
```
Similar, redundant if

EugeneTheDev commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/src/jvmMain/kotlin/ai/koog/prompt/executor/clients/bedrock/BedrockLLMClient.kt:
```diff
@@ -0,0 +1,292 @@
+                }
+            } catch (e: Exception) {
+                logger.warn(e) { "Failed to parse Bedrock stream chunk: $chunkJsonString" }
+                ""
```
Let's rethrow an exception explicitly here instead of sliently swallowing it

EugeneTheDev commented on prompt/prompt-executor/prompt-executor-llms-all/src/commonMain/kotlin/ai/koog/prompt/executor/llms/all/SimplePromptExecutors.kt:
```diff
@@ -44,4 +47,18 @@ public fun simpleGoogleAIExecutor(apiKey: String): SingleLLMPromptExecutor = Sin
+public fun simpleBedrockExecutor(
+    awsAccessKeyId: String,
+    awsSecretAccessKey: String,
+    settings: BedrockClientSettings = BedrockClientSettings()
```
This doesn't compile, Bedrock client has to be added to the module dependencies

kaeawc replied:
It is - the issue is that Bedrock is in `jvmMain` and this code is in `commonMain`. Going to think about it a bit more before changing

EugeneTheDev commented on prompt/prompt-executor/prompt-executor-clients/src/commonMain/kotlin/ai/koog/prompt/executor/clients/InternalAPI.kt:
```diff
@@ -0,0 +1,17 @@
+    AnnotationTarget.PROPERTY_SETTER,
+    AnnotationTarget.PROPERTY_SETTER
+)
+public annotation class InternalAPI
```
Let's rename it to `InternalLLMClientApi` to better highlight the scope and purpose. We already have `InternalAgentToolsApi` annotation, so let's continue with this scoped annotations approach for now. We can introduce some single global `InternalKoogAPI` annotation later.

Note for maintainers: we will need to update custom inspection `MissingKDocForPublicAPI.inspection.kts` so that it doesn't require KDocs on public symbols marked with this annotation.

EugeneTheDev commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-bedrock-client/build.gradle.kts:
```diff
@@ -0,0 +1,60 @@
+            dependencies {
+                api(libs.ktor.client.cio)
+                implementation(project(":prompt:prompt-executor:prompt-executor-clients:prompt-executor-anthropic-client"))
+                implementation(project(":prompt:prompt-executor:prompt-executor-clients:prompt-executor-google-client"))
```
Do we need dependencies on OpenAI and Google clients here? Looks like only Anthropic DTOs are reused for now, please consider removing these dependencies

EugeneTheDev commented on examples/src/main/kotlin/ai/koog/agents/example/client/BedrockAgent.kt:
```diff
@@ -0,0 +1,48 @@
+import ai.koog.prompt.executor.llms.all.simpleBedrockExecutor
+import kotlinx.coroutines.runBlocking
+
+fun main() = runBlocking {
```
nit: Please specify return type (`Unit`) explicitly, otherwise there's a warning about potential non-Unit return type in main (since `runBlocking` can return some value).

EugeneTheDev reviewed (approved):
Generally looks good, thank you for this PR. I left a few minor comments which should be easy to fix, mostly style issues and couple small fixes, please check these. And also, please mark all public symbols in the Bedrock client module as `internal`, except for these that are actually meant to be public (client, config, models list).
